planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+)
+
+type outerJoinEliminator struct {
+	hasDistinct int
+	cols        [][]*expression.Column
+	schemas     []*expression.Schema
+}
+
+// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
+// 1. outer join elimination: For example left outer join, if the parent only use the
+//    columns from left table and the join key of right table(the inner table) is a unique
+//    key of the right table. the left outer join can be eliminated.
+// 2. outer join elimination with distinct: For example left outer join. If the parent
+//    only use the columns from left table with 'distinct' label. The left outer join can
+//    be eliminated.
+func (o *outerJoinEliminator) tryToEliminateOuterJoin(p *LogicalJoin) LogicalPlan {
+	if len(p.children) != 2 {
+		return nil
+	}
+	switch p.JoinType {
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, isLeft int) LogicalPlan {
+	// outer join elimination with distinct
+	if o.hasDistinct > 0 {
+		cols := o.cols[o.hasDistinct-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			if !p.children[1^isLeft].Schema().Contains(col) {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^isLeft]
+		}
+	}
+
+	// outer join elimination without distinct
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) < 1 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := p.children[1^isLeft].Schema().FindColumn(columnName); c == nil {
+			return nil
+		}
+	}
+	// second, check whether the other side join keys are unique keys
+	for _, keyInfo := range p.children[isLeft].Schema().Keys {
+		keyInfoContainAllCols := true
+		var joinKeys []*expression.Column
+		if isLeft > 0 {
+			joinKeys = p.RightJoinKeys
+		} else {
+			joinKeys = p.LeftJoinKeys
+		}
+		for _, col := range joinKeys {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := expression.NewSchema(keyInfo...).FindColumn(columnName); c == nil {
```

Comment:
I think we should check whether a unique key can be covered by the join key. But here we actually check whether the join key can be covered by the unique key, it's not correct. Take this as an example:
```sql
create table t1(a bigint, b bigint, c bigint, unique key idx(a, b));
create table t2(a bigint, b bigint, c bigint);

-- This outer join can be eliminated in this patch. But it should not actually.
select t2.c from t2 left join t1 on t2.a = t1.a;

-- This outer join can not be eliminated in this patch. But it should actually.
select t2.c from t2 left join t1 on t2.a = t1.a and t2.b = t1.b and t2.c = t1.c;
```

Reply:
Why the first query can eliminate join? Consider the following situation:
```
table t1:
+------+------+------+
| a    | b    | c    |
+------+------+------+
|    1 |    1 |    1 |
|    1 |    2 |    1 |
|    2 |    2 |    1 |
+------+------+------+
table t2:
+------+------+------+
| a    | b    | c    |
+------+------+------+
|    1 |    1 |    1 |
+------+------+------+
```
If we don't eliminate join the answer is:
```
+------+
| c    |
+------+
|    1 |
|    1 |
+------+
```
But we eliminate it will get:
```
+------+
| c    |
+------+
|    1 |
+------+
```

Reply:
@lzmhhh123 No, I mean the outer join in the first query should not be eliminated. But the outer join in the second query should be eliminated.

In your patch, both the two queries can not be eliminated:
```sql
TiDB(localhost:4000) > drop table if exists t1, t2;
Query OK, 0 rows affected (0.03 sec)

TiDB(localhost:4000) > create table t1(a bigint, b bigint, c bigint, unique key idx(a, b));
Query OK, 0 rows affected (0.01 sec)

TiDB(localhost:4000) > create table t2(a bigint, b bigint, c bigint);
Query OK, 0 rows affected (0.02 sec)

TiDB(localhost:4000) > desc select t2.c from t2 left join t1 on t2.a = t1.a;
+--------------------------+----------+------+-------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                           |
+--------------------------+----------+------+-------------------------------------------------------------------------+
| Projection_5             | 12500.00 | root | test.t2.c                                                               |
| └─HashLeftJoin_9         | 12500.00 | root | left outer join, inner:TableReader_13, equal:[eq(test.t2.a, test.t1.a)] |
|   ├─TableReader_11       | 10000.00 | root | data:TableScan_10                                                       |
|   │ └─TableScan_10       | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo             |
|   └─TableReader_13       | 10000.00 | root | data:TableScan_12                                                       |
|     └─TableScan_12       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo             |
+--------------------------+----------+------+-------------------------------------------------------------------------+
6 rows in set (0.01 sec)

TiDB(localhost:4000) > desc select t2.c from t2 left join t1 on t2.a = t1.a and t2.b = t1.b and t2.c = t1.c;
+--------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                                                                             |
+--------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------+
| Projection_5             | 12500.00 | root | test.t2.c                                                                                                                 |
| └─HashLeftJoin_10        | 12500.00 | root | left outer join, inner:TableReader_14, equal:[eq(test.t2.a, test.t1.a) eq(test.t2.b, test.t1.b) eq(test.t2.c, test.t1.c)] |
|   ├─TableReader_12       | 10000.00 | root | data:TableScan_11                                                                                                         |
|   │ └─TableScan_11       | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
|   └─TableReader_14       | 10000.00 | root | data:TableScan_13                                                                                                         |
|     └─TableScan_13       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
+--------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```